graphex.hpp

```diff
@@ -309,63 +293,102 @@ class GraphEx {
             }
             return false;
         };
-        for (auto& inputNode : _inputNodes) {
-            if (vis.find(inputNode) == vis.end())
-                dfs(inputNode);
+        for (auto& node : _nodes) {
+            if (vis.find(node.get()) == vis.end())
+                dfs(node.get());
         }
     }
 
     /// @brief run the graph execution from input nodes
     void execute()
     {
-        // create thread _pool with 4 worker threads
-        std::unordered_set<BaseNode*> processed;
-        std::queue<BaseNode*> qu;
-
-        for (auto& v : _inputNodes) {
-            qu.emplace(v);
-            processed.emplace(v);
-            for (auto& k : v->_nextNodes)
-                k->parentEnqueued();
-        }
-
-        while (!qu.empty()) {
-            decltype(auto) nxt = qu.front();
-            qu.pop();
-            _pool.push(std::bind(&BaseNode::execute, nxt));
-            for (auto& nextNode : nxt->_nextNodes) {
-                if (processed.find(nextNode) == processed.end() &&
-                    nextNode->allParentEnqueued()) {
-                    qu.emplace(nextNode);
-                    processed.emplace(nextNode);
-                    for (auto& v : nextNode->_nextNodes)
-                        v->parentEnqueued();
-                }
-            }
-        }
+        for (auto& nodePtr : _nodes)
+            if (!nodePtr.get()->getPendingCount())
+                _pool.push(std::bind(&BaseNode::execute, nodePtr.get()));
+        while (_finishedCount != _nodes.size()) // TODO: replace with blocking instead of polling?
```

Comment:
Might need to reset finished count in the reset function otherwise the graph will run forever after reset since _finishedCount will be > _nodes.size()

Reply:
Good point, t miss mat cho nay. Cai nay ko fix duoc tai sao cai `BM_GraphEX` cua toi cu run forever though, ko hieu co chuyen gi.